server/src/main/java/io/crate/expression/scalar/cast/CastFunction.java

```diff
@@ -55,45 +56,80 @@ public static void register(ScalarFunctionModule module) {
+                        throw new ConversionException(argument, returnType);
+                    },
+                    (argument, returnType) -> {
+                        throw new ConversionException(argument, returnType);
+                    }
+                );
+            }
         );
         module.register(
             Signature
-                .scalar(TRY_CAST_NAME,
-                        parseTypeSignature("E"),
-                        parseTypeSignature("V"),
-                        parseTypeSignature("V"))
+                .scalar(
+                    TRY_CAST_NAME,
+                    parseTypeSignature("E"),
+                    parseTypeSignature("V"),
+                    parseTypeSignature("V"))
                 .withTypeVariableConstraints(typeVariable("E"), typeVariable("V")),
-            (signature, args) -> new CastFunction(
-                new FunctionInfo(new FunctionIdent(TRY_CAST_NAME, args), args.get(1)),
-                signature,
-                (argument, returnType) -> Literal.NULL,
-                (argument, returnType) -> null
-            )
+            (signature, args) -> {
+                var targetType = args.get(1);
+                return new CastFunction(
+                    new FunctionInfo(new FunctionIdent(TRY_CAST_NAME, args), targetType),
+                    targetType::explicitCast,
+                    signature,
+                    (argument, returnType) -> Literal.NULL,
+                    (argument, returnType) -> null
+                );
+            }
+        );
+        module.register(
+            Signature
+                .scalar(
+                    IMPLICIT_CAST_NAME,
+                    parseTypeSignature("E"),
+                    parseTypeSignature("V"),
+                    parseTypeSignature("V"))
+                .withTypeVariableConstraints(typeVariable("E"), typeVariable("V")),
+            (signature, args) -> {
+                var targetType = args.get(1);
+                return new CastFunction(
+                    new FunctionInfo(new FunctionIdent(IMPLICIT_CAST_NAME, args), targetType),
+                    targetType::implicitCast,
+                    signature,
+                    (argument, returnType) -> {
+                        throw new ConversionException(argument, returnType);
+                    },
+                    (argument, returnType) -> {
+                        throw new ConversionException(argument, returnType);
+                    }
+                );
+            }
         );
     }
 
 
     private final DataType<?> returnType;
+    private final Function<Object, Object> cast;
     private final FunctionInfo info;
     private final Signature signature;
     private final BiFunction<Symbol, DataType<?>, Symbol> onNormalizeException;
     private final BiFunction<Object, DataType<?>, Object> onEvaluateException;
 
     private CastFunction(FunctionInfo info,
+                         Function<Object, Object> cast,
```

Comment:
I wonder if we should make 3 concrete implementations instead.  At this point pretty much everything this function does is abstracted away.  WDYT?

Reply:
yes, I thought of splitting it, but was dragged away by cast function printing stuff.